File: src/GeoTools/GeoPoint.cpp
//
// Created by pshivaraman on 2021-08-19.
//

#define _USE_MATH_DEFINES

#include <cmath>

#include "GeoTools/GeoPoint.h"
#include <limits>
#include <memory>
#include "MathTools/MathUtil.h"
#include "GeoTools/GeoUtil.h"

using namespace AviationCalcUtil::MathTools;
using namespace AviationCalcUtil::GeoTools;

GeoPoint::GeoPoint(const GeoPoint &point) {
    lat = point.getLat();
    lon = point.getLon();
    alt = point.getAlt();
}

GeoPoint::GeoPoint(double lat, double lon, double alt) {
    this->lat = std::fmin(std::fmax(lat, -90), 90);
    this->lon = GeoUtil::normalizeLongitude(lon);
    this->alt = alt;
}

void GeoPoint::moveByM(double bearing, double distance) {
    double R = GeoUtil::EARTH_RADIUS_M + MathUtil::convertFeetToMeters(alt);
    double bearingRads = MathUtil::convertDegreesToRadians(GeoUtil::normalizeHeading(bearing));
    double lat1 = MathUtil::convertDegreesToRadians(lat);
    double lon1 = MathUtil::convertDegreesToRadians(lon);

    double lat2 = std::asin((std::sin(lat1) * std::cos(distance / R) +
                             std::cos(lat1) * std::sin(distance / R) * std::cos(bearingRads)));
    double lon2 = lon1 + std::atan2(std::sin(bearingRads) * std::sin(distance / R) * std::cos(lat1),
                                    std::cos(distance / R) - std::sin(lat1) * std::sin(lat2));

    setLat(MathUtil::convertRadiansToDegrees(lat2));
    setLon(MathUtil::convertRadiansToDegrees(lon2));
}

void GeoPoint::moveByNMi(double bearing, double distance) {
    moveByM(bearing, MathUtil::convertNauticalMilesToMeters(distance));
}

double GeoPoint::flatDistanceM(const GeoPoint &point1, const GeoPoint &point2) {
    double phi1 = MathUtil::convertDegreesToRadians(point1.getLat());
    double phi2 = MathUtil::convertDegreesToRadians(point2.getLat());
    double deltaPhi = MathUtil::convertDegreesToRadians(point2.getLat() - point1.getLat());
    double deltaLambda = MathUtil::convertDegreesToRadians(point2.getLon() - point1.getLon());

    double a = std::pow(std::sin(deltaPhi / 2), 2) +
               std::cos(phi1) * std::cos(phi2) *
               std::pow(std::sin(deltaLambda / 2), 2);

    double c = 2 * std::atan2(std::sqrt(a), std::sqrt(1 - a));

    double d = GeoUtil::EARTH_RADIUS_M * c;

    return d;
}

double GeoPoint::flatDistanceNMi(const GeoPoint &point1, const GeoPoint &point2) {
    return MathUtil::convertMetersToNauticalMiles(flatDistanceM(point1, point2));
}

double GeoPoint::distanceM(const GeoPoint &point1, const GeoPoint &point2) {
    double flatDist = flatDistanceM(point1, point2);
    double altDist = MathUtil::convertFeetToMeters(std::fabs(point2.getAlt() - point1.getAlt()));

    return std::sqrt(std::pow(flatDist, 2) + std::pow(altDist, 2));
}

double GeoPoint::distanceNMi(const GeoPoint &point1, const GeoPoint &point2) {
    return MathUtil::convertMetersToNauticalMiles(distanceM(point1, point2));
}

std::unique_ptr<GeoPoint>
GeoPoint::intersection(const GeoPoint &point1, double bearing1, const GeoPoint &point2, double bearing2) {
    // Conversions to radians
    double phi1 = MathUtil::convertDegreesToRadians(point1.getLat());
    double phi2 = MathUtil::convertDegreesToRadians(point2.getLat());
    double lambda1 = MathUtil::convertDegreesToRadians(point1.getLon());
    double lambda2 = MathUtil::convertDegreesToRadians(point2.getLon());
    double theta13 = MathUtil::convertDegreesToRadians(bearing1);
    double theta23 = MathUtil::convertDegreesToRadians(bearing2);
    double deltaPhi = phi2 - phi1;
    double deltaLambda = lambda2 - lambda1;

    // Angular distance (lat1, lon1) - (lat2, lon2)
    double sigma12 = 2 * std::asin(std::sqrt(std::pow(std::sin(deltaPhi / 2), 2)
                                             +
                                             std::cos(phi1) * std::cos(phi2) * std::pow(std::sin(deltaLambda / 2), 2)));

    // Coincident points
    if (sigma12 < std::numeric_limits<double>::epsilon()) {
        return std::make_unique<GeoPoint>(point1);
    }

    // Initial/Final Bearing between points
    double cosThetaA = (std::sin(phi2) - std::sin(phi1) * std::cos(sigma12)) / (std::sin(sigma12) * std::cos(phi1));
    double cosThetaB = (std::sin(phi1) - std::sin(phi2) * std::cos(sigma12)) / (std::sin(sigma12) * std::cos(phi2));
    double thetaA = std::acos(std::fmin(std::fmax(cosThetaA, -1), 1)); // Prevent rounding errors
    double thetaB = std::acos(std::fmin(std::fmax(cosThetaB, -1), 1)); // Prevent rounding errors

    double theta12 = std::sin(deltaLambda) > 0 ? thetaA : 2 * M_PI - thetaA;
    double theta21 = std::sin(deltaLambda) > 0 ? 2 * M_PI - thetaB : thetaB;

    double alpha1 = theta13 - theta12;// Angle 2-1-3
    double alpha2 = theta21 - theta23;// Angle 1-2-3

    // Infinite intersections
    if (std::sin(alpha1) == 0 && std::sin(alpha2) == 0) {
        return nullptr;
    }

    // Ambiguous Intersection (antipodal?)
    if (std::sin(alpha1) * std::sin(alpha2) < 0) {
        return nullptr;
    }

    double cosAlpha3 = -std::cos(alpha1) * std::cos(alpha2) + std::sin(alpha1) * std::sin(alpha2) * std::cos(sigma12);

    double sigma13 = std::atan2(std::sin(sigma12) * std::sin(alpha1) * std::sin(alpha2),
                                std::cos(alpha2) + std::cos(alpha1) * cosAlpha3);

    double phi3 = std::asin(std::fmin(
            std::fmax(std::sin(phi1) * std::cos(sigma13) + std::cos(phi1) * std::sin(sigma13) * std::cos(theta13), -1),
            1));

    double deltaLambda13 = std::atan2(std::sin(theta13) * std::sin(sigma13) * std::cos(phi1),
                                      std::cos(sigma13) - std::sin(phi1) * std::sin(phi3));

    double lambda3 = lambda1 + deltaLambda13;

    return std::make_unique<GeoPoint>(MathUtil::convertRadiansToDegrees(phi3),
                                      MathUtil::convertRadiansToDegrees(lambda3));
}

double GeoPoint::initialBearing(const GeoPoint &point1, const GeoPoint &point2) {
    // Convert to Radians
    double phi1 = MathUtil::convertDegreesToRadians(point1.getLat());
    double phi2 = MathUtil::convertDegreesToRadians(point2.getLat());
    double lambda1 = MathUtil::convertDegreesToRadians(point1.getLon());
    double lambda2 = MathUtil::convertDegreesToRadians(point2.getLon());

    // Find angle between the two
    double y = std::sin(lambda2 - lambda1) * std::cos(phi2);
    double x = std::cos(phi1) * std::sin(phi2) - std::sin(phi1) * std::cos(phi2) * std::cos(lambda2 - lambda1);

    double theta = std::atan2(y, x);

    // Convert from -180, +180 to 0, 359
    return GeoUtil::normalizeHeading(MathUtil::convertRadiansToDegrees(theta));
}

double GeoPoint::finalBearing(const GeoPoint &point1, const GeoPoint &point2) {
    // Calculate initial bearing from end to start and reverse
    return std::fmod(initialBearing(point2, point1) + 180, 360);
}

bool GeoPoint::equals(const GeoPoint &o) const {
    return lat == o.getLat() && lon == o.getLon();
}

double GeoPoint::getLat() const {
    return lat;
}

void GeoPoint::setLat(double newLat) {
    lat = std::fmin(std::fmax(newLat, -90), 90);
}

double GeoPoint::getLon() const {
    return lon;
}

void GeoPoint::setLon(double newLon) {
    lon = GeoUtil::normalizeLongitude(newLon);
}

double GeoPoint::getAlt() const {
    return alt;
}

void GeoPoint::setAlt(double newAlt) {
    alt = newAlt;
}

double GeoPoint::operator-(const GeoPoint &point) const {
    return distanceM(*this, point);
}

/*
AviationCalcUtil::GeoTools::GeoPoint *CreateGeoPoint(double lat, double lon, double alt) {
    return new GeoPoint(lat, lon, alt);
}

AviationCalcUtil::GeoTools::GeoPoint *CopyGeoPoint(AviationCalcUtil::GeoTools::GeoPoint* ptr) {
    return new AviationCalcUtil::GeoTools::GeoPoint(*ptr);
}

void DisposeGeoPoint(AviationCalcUtil::GeoTools::GeoPoint *ptr) {
    if (ptr != NULL) {
        delete ptr;
        ptr = NULL;
    }
}

void GeoPointMoveByM(AviationCalcUtil::GeoTools::GeoPoint *ptr, double bearing, double distance) {
    if (ptr != NULL) {
        ptr->moveByM(bearing, distance);
    }
}

void GeoPointMoveByNMi(AviationCalcUtil::GeoTools::GeoPoint *ptr, double bearing, double distance)
{
    if (ptr != NULL) {
        ptr->moveByNMi(bearing, distance);
    }
}

double GeoPointFlatDistanceM(AviationCalcUtil::GeoTools::GeoPoint *ptr1, AviationCalcUtil::GeoTools::GeoPoint *ptr2) {
    if (ptr1 != NULL && ptr2 != NULL)
    {
        return GeoPoint::flatDistanceM(*ptr1, *ptr2);
    }
    return -1;
}

double GeoPointFlatDistanceNMi(AviationCalcUtil::GeoTools::GeoPoint *ptr1, AviationCalcUtil::GeoTools::GeoPoint *ptr2) {
    if (ptr1 != NULL && ptr2 != NULL)
    {
        return GeoPoint::flatDistanceNMi(*ptr1, *ptr2);
    }
    return -1;
}

double GeoPointDistanceM(AviationCalcUtil::GeoTools::GeoPoint *ptr1, AviationCalcUtil::GeoTools::GeoPoint *ptr2) {
    if (ptr1 != NULL && ptr2 != NULL)
    {
        return GeoPoint::distanceM(*ptr1, *ptr2);
    }
    return -1;
}

double GeoPointDistanceNMi(AviationCalcUtil::GeoTools::GeoPoint *ptr1, AviationCalcUtil::GeoTools::GeoPoint *ptr2) {
    if (ptr1 != NULL && ptr2 != NULL)
    {
        return GeoPoint::distanceNMi(*ptr1, *ptr2);
    }
    return -1;
}

AviationCalcUtil::GeoTools::GeoPoint*
GeoPointIntersection(AviationCalcUtil::GeoTools::GeoPoint *ptr1, double bearing1, AviationCalcUtil::GeoTools::GeoPoint *ptr2,
             double bearing2) {
    if (ptr1 != NULL && ptr2 != NULL)
    {
        return GeoPoint::intersection(*ptr1, bearing1, *ptr2, bearing2).release();
    }
    return NULL;
}

double GeoPointInitialBearing(AviationCalcUtil::GeoTools::GeoPoint *ptr1, AviationCalcUtil::GeoTools::GeoPoint *ptr2) {
    if (ptr1 != NULL && ptr2 != NULL)
    {
        return GeoPoint::initialBearing(*ptr1, *ptr2);
    }
    return -1;
}

double GeoPointFinalBearing(AviationCalcUtil::GeoTools::GeoPoint *ptr1, AviationCalcUtil::GeoTools::GeoPoint *ptr2) {
    if (ptr1 != NULL && ptr2 != NULL)
    {
        return GeoPoint::finalBearing(*ptr1, *ptr2);
    }
    return -1;
}

bool GeoPointEquals(AviationCalcUtil::GeoTools::GeoPoint *ptr1, AviationCalcUtil::GeoTools::GeoPoint *ptr2) {
    if (ptr1 != NULL && ptr2 != NULL)
    {
        return ptr1->equals(*ptr2);
    }
    return false;
}

double GeoPointGetLat(AviationCalcUtil::GeoTools::GeoPoint *ptr) {
    if (ptr != NULL)
    {
        return ptr->getLat();
    }
    return -1;
}

void GeoPointSetLat(AviationCalcUtil::GeoTools::GeoPoint *ptr, double newLat) {
    if (ptr != NULL)
    {
        ptr->setLat(newLat);
    }
}

double GeoPointGetLon(AviationCalcUtil::GeoTools::GeoPoint *ptr) {
    if (ptr != NULL)
    {
        return ptr->getLon();
    }
    return -1;
}

void GeoPointSetLon(AviationCalcUtil::GeoTools::GeoPoint *ptr, double newLon) {
    if (ptr != NULL)
    {
        ptr->setLon(newLon);
    }
}

double GeoPointGetAlt(AviationCalcUtil::GeoTools::GeoPoint *ptr) {
    if (ptr != NULL)
    {
        return ptr->getAlt();
    }
    return -1;
}

void GeoPointSetAlt(AviationCalcUtil::GeoTools::GeoPoint *ptr, double newAlt) {
    if (ptr != NULL)
    {
        ptr->setAlt(newAlt);
    }
}*/

File: src/geo_tools/geo_point.rs
use std::{f64::consts::PI, ops::Sub};

use crate::units::{angle::Angle, length::Length};

use super::{bearing::Bearing, geo_util, latitude::Latitude, longitude::Longitude};

/// Represents a 3D Coordinate Point on the globe.
#[derive(Clone, Copy, Default, PartialEq, Debug)]
pub struct GeoPoint {
    pub lat: Latitude,
    pub lon: Longitude,
    pub alt: Length,
}

impl Sub for GeoPoint {
    type Output = Length;

    fn sub(self, rhs: Self) -> Self::Output {
        return self.distance(&rhs);
    }
}

impl GeoPoint {
    /// Creates a new GeoPoint
    pub fn new(lat: Latitude, lon: Longitude, alt: Length) -> GeoPoint {
        GeoPoint { lat, lon, alt }
    }

    /// Moves the GeoPoint on a bearing by a distance
    pub fn move_by(&mut self, bearing: Bearing, distance: Length) {
        let r = geo_util::EARTH_RADIUS + self.alt;
        let bearing_rads = bearing.as_radians();
        let lat1 = self.lat.as_radians();
        let lon1 = self.lon.as_radians();

        let lat2 = f64::asin(
            f64::sin(lat1) * f64::cos((distance / r).as_meters()) + f64::cos(lat1) * f64::sin((distance / r).as_meters()) * f64::cos(bearing_rads),
        );

        let lon2 = lon1
            + f64::atan2(
                f64::sin(bearing_rads) * f64::sin((distance / r).as_meters()) * f64::cos(lat1),
                f64::cos((distance / r).as_meters()) - f64::sin(lat1) * f64::sin(lat2),
            );

        self.lat = Latitude(Angle(lat2));
        self.lon = Longitude(Angle(lon2));
    }

    /// Calculates the flat distance between two GeoPoints
    pub fn flat_distance(&self, other: &GeoPoint) -> Length {
        let phi1 = self.lat.as_radians();
        let phi2 = other.lat.as_radians();
        let delta_phi = other.lat.as_radians() - self.lat.as_radians();
        let delta_lambda = other.lon.as_radians() - self.lon.as_radians();

        let a = f64::sin(delta_phi / 2.0).powi(2) + f64::cos(phi1) * f64::cos(phi2) * f64::sin(delta_lambda / 2.0).powi(2);

        // Check for opposite end of planet
        if (1.0 - a <= 0.0) {
            // Return half circumference
            return geo_util::EARTH_RADIUS * PI;
        }

        let c = 2.0 * f64::atan2(f64::sqrt(a), f64::sqrt(1.0 - a));

        let d = geo_util::EARTH_RADIUS * c;

        return d;
    }

    /// Calculates the distance including altitude between two GeoPoints
    pub fn distance(&self, other: &GeoPoint) -> Length {
        let flat_dist = self.flat_distance(other);
        let alt_dist = (other.alt - self.alt).abs();

        return Length::sqrt(flat_dist.powi(2) + alt_dist.powi(2));
    }

    /// Calculates the intersection between two points and two courses
    /// 
    /// *Note:* This function will not account for reciprocal of the bearings
    pub fn intersection(&self, bearing_1: Bearing, other: &GeoPoint, bearing_2: Bearing) -> Option<GeoPoint> {
        // Conversions to radians
        let phi_1 = self.lat.as_radians();
        let phi_2 = other.lat.as_radians();
        let lambda_1 = self.lon.as_radians();
        let lambda_2 = other.lon.as_radians();
        let theta_1_3 = bearing_1.as_radians();
        let theta_2_3 = bearing_2.as_radians();
        let delta_phi = phi_2 - phi_1;
        let delta_lambda = lambda_2 - lambda_1;

        // Angular distance (lat1, lon1) - (lat2, lon2)
        let sigma_1_2 = 2.0
            * f64::asin(f64::sqrt(
                f64::sin(delta_phi / 2.0).powi(2) + f64::cos(phi_1) * f64::cos(phi_2) * f64::sin(delta_lambda / 2.0).powi(2),
            ));

        // Coincident Points
        if (sigma_1_2 < f64::EPSILON) {
            return Some(self.clone());
        }

        // Initial/Final bearing between points
        let cos_theta_a = (f64::sin(phi_2) - f64::sin(phi_1) * f64::cos(sigma_1_2)) / (f64::sin(sigma_1_2) * f64::cos(phi_1));
        let cos_theta_b = (f64::sin(phi_1) - f64::sin(phi_2) * f64::cos(sigma_1_2)) / (f64::sin(sigma_1_2) * f64::cos(phi_2));
        let theta_a = f64::acos(f64::min(f64::max(cos_theta_a, -1.0), 1.0)); // Prevent rounding errors
        let theta_b = f64::acos(f64::min(f64::max(cos_theta_b, -1.0), 1.0)); // Prevent rounding errors

        let theta_1_2 = if f64::sin(delta_lambda) > 0.0 { theta_a } else { 2.0 * PI - theta_a };
        let theta_2_1 = if f64::sin(delta_lambda) > 0.0 { 2.0 * PI - theta_b } else { theta_b };

        let alpha_1 = theta_1_3 - theta_1_2; // Angle 2-1-3
        let alpha_2 = theta_2_1 - theta_2_3; // Angle 1-2-3

        // Infinite intersections
        if (f64::sin(alpha_1) == 0.0 && f64::sin(alpha_2) == 0.0) {
            return None;
        }

        // Ambiguous Intersection (antipodal?)
        if (f64::sin(alpha_1) * f64::sin(alpha_2) < 0.0) {
            return None;
        }

        let cos_alpha_3 = -f64::cos(alpha_1) * f64::cos(alpha_2) + f64::sin(alpha_1) * f64::sin(alpha_2) * f64::cos(sigma_1_2);

        let sigma_1_3 = f64::atan2(
            f64::sin(sigma_1_2) * f64::sin(alpha_1) * f64::sin(alpha_2),
            f64::cos(alpha_2) + f64::cos(alpha_1) * cos_alpha_3,
        );

        let phi_3 = f64::asin(f64::min(
            f64::max(
                f64::sin(phi_1) * f64::cos(sigma_1_3) + f64::cos(phi_1) * f64::sin(sigma_1_3) * f64::cos(theta_1_3),
                -1.0,
            ),
            1.0,
        ));

        let delta_lambda_1_3 = f64::atan2(
            f64::sin(theta_1_3) * f64::sin(sigma_1_3) * f64::cos(phi_1),
            f64::cos(sigma_1_3) - f64::sin(phi_1) * f64::sin(phi_3),
        );

        let lambda_3 = lambda_1 + delta_lambda_1_3;

        return Some(GeoPoint::new(Latitude(Angle(phi_3)), Longitude(Angle(lambda_3)), Length(0.0)));
    }

    /// Find closest intersection between two points and two radials.
    /// 
    /// This function will try both the radial and the reciprocal
    pub fn closest_intersection(&self, radial_1: Bearing, other: &GeoPoint, radial_2: Bearing) -> Option<GeoPoint> {
        // Try all possible radials
        let mut intersections = Vec::new();
        intersections.push(GeoPoint::intersection(&self, radial_1, other, radial_2));
        intersections.push(GeoPoint::intersection(&self, radial_1, other, -radial_2));
        intersections.push(GeoPoint::intersection(&self, -radial_1, other, radial_2));
        intersections.push(GeoPoint::intersection(&self, -radial_1, other, -radial_2));

        let mut closest_dist = Length(f64::MAX);
        let mut closest_intersection = None;

        for (_i, arg) in intersections.iter().enumerate() {
            if let Some(this_intersection) = arg {
                let dist = self.flat_distance(&this_intersection);

                if dist < closest_dist {
                    closest_intersection = Some(this_intersection.to_owned());
                    closest_dist = dist;
                }
            }
        }

        return closest_intersection;
    }

    /// Calculates initial bearing from one GeoPoint to other
    pub fn initial_bearing(&self, other: &GeoPoint) -> Bearing {
        // Convert to Radians
        let phi_1 = self.lat.as_radians();
        let phi_2 = other.lat.as_radians();
        let lambda_1 = self.lon.as_radians();
        let lambda_2 = self.lon.as_radians();

        // Find angle between the two
        let y = f64::sin(lambda_2 - lambda_1) * f64::cos(phi_2);
        let x = f64::cos(phi_1) * f64::sin(phi_2) - f64::sin(phi_1) * f64::cos(phi_2) * f64::cos(lambda_2 - lambda_1);

        let theta = f64::atan2(y, x);

        // Convert from -180, +180 to 0, 359
        return Bearing(Angle(theta));
    }

    /// Calculates final bearing from one GeoPoint to other
    pub fn final_bearing(&self, other: &GeoPoint) -> Bearing {
        // Calculate initial bearing from end to start and reverse
        return -other.initial_bearing(self);
    }
}
